### ISSUE TITLE
Addon-Docs: Changing the theme via api doesn't persist in Docs

### ISSUE THREAD

**Issue by angelolucas** [storybookjs/storybook#10523]
**Describe the bug**
I have an addon tool that handles the storybook theme through the api:
```
// inside my addon

api.setOptions ({
  theme: themes[base],
});
```
The Storybook matches, but Docs don't change.
Docs works with the dark theme just by adding through addParametes:
```
// config.js

import { themes } from '@storybook/theming';

addParameters({
  options: {
    theme: themes.dark,
},
```

**Expected behavior**
That when changing the theme through the storybook API, it trigger the Docs theme

**Screenshots**
![issue](https://user-images.githubusercontent.com/828776/80021314-affe2380-84b0-11ea-85d9-1c334ff79000.png)

**System:**
Environment Info:

  System:
    OS: macOS 10.15.4
    CPU: (4) x64 Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
  Binaries:
    Node: 10.16.0 - /usr/local/bin/node
    Yarn: 1.9.4 - /usr/local/bin/yarn
    npm: 6.13.7 - /usr/local/bin/npm
  Browsers:
    Chrome: 81.0.4044.122
    Firefox: 74.0
    Safari: 13.1
  npmPackages:
    @storybook/addon-a11y: ^5.3.18 => 5.3.18
    @storybook/addon-actions: ^5.3.18 => 5.3.18
    @storybook/addon-docs: ^5.3.18 => 5.3.18
    @storybook/addon-knobs: ^5.3.18 => 5.3.18
    @storybook/addon-storysource: ^5.3.18 => 5.3.18
    @storybook/addon-viewport: ^5.3.18 => 5.3.18
    @storybook/addons: ^5.3.18 => 5.3.18
    @storybook/react: ^5.3.18 => 5.3.18
    @storybook/storybook-deployer: ^2.8.3 => 2.8.5

**Comment by shilman** [comment]
In 6.0 we've moved to a parallel theming structure for docs:

https://github.com/storybookjs/storybook/blob/next/addons/docs/docs/theming.md#storybook-theming

**Comment by angelolucas** [comment]
@shilman this instruction set the theme in the initial state. Is there a way to change at run time?

I want to change the theme of the Storybook, Docs and Design System at once:

<img width="700" alt="Screen Shot 2020-04-23 at 15 59 09" src="https://user-images.githubusercontent.com/828776/80138898-f2d8fd80-857b-11ea-867f-6cf23725376d.png">

**Comment by shilman** [comment]
I believe docs is dynamically themable but Storybook UI is not cc @ndelangen

**Comment by hipstersmoothie** [comment]
I'm looking for this same thing too for https://github.com/hipstersmoothie/storybook-dark-mode/. I need a way to dynamically set the theme for the `adon-docs` tab.

**Comment by shilman** [comment]
@hipstersmoothie OK I misunderstood the issue. In `master`, docs is themed by the `options.theme` parameter per this issue. In `next`, docs is themed via the `docs.theme` parameter per https://github.com/storybookjs/storybook/pull/10114. How did your addon work before `api.setOptions` existed?

@tmeasday is it possible to update parameters dynamically, or perhaps we should consider moving to `globalArgs` instead?

**Comment by ndelangen** [comment]
setting the theme is possible at runtime for the manager:
https://github.com/storybookjs/storybook/blob/a14bf2aa77802bfd53721246ac0879ee2d85ac88/lib/api/src/modules/layout.ts#L223-L262

**Comment by hipstersmoothie** [comment]
@shilman This is how I have always set the theme for storybook https://github.com/hipstersmoothie/storybook-dark-mode/blob/master/src/Tool.tsx#L90 I'm not sure that theming ever worked with the docs tab (https://github.com/hipstersmoothie/storybook-dark-mode/issues/70) for this addon.

> is it possible to update parameters dynamically

This is exactly what I need. I saw the globalArg function but it seems like it was for the new `args` feature

**Comment by tmeasday** [comment]
> is it possible to update parameters dynamically

No? I haven't really been following but the obvious answer to that question is of course not. That's exactly why we made global args.

**Comment by hipstersmoothie** [comment]
So the question remains: How can I dynamically theme `addon-docs`? If it's through global args that's cool. I just read through the doc and it doesn't mention parameters very much.

**Comment by shilman** [comment]
@tmeasday in that case either #10114 was wrong and we should switch to `globalArgs` (which is possible because we haven't released 6.0 yet), or we should forego dynamic theming in docs. thoughts?

**Comment by ndelangen** [comment]
I think it should use global args then?

**Comment by tmeasday** [comment]
Yeah, I guess?

I think we’ll likely keep running into challenges deciding if things should be configured via parameters vs global args. Perhaps we should make a pattern where certain things are can be configured both ways.

What do you think folks?

**Comment by ndelangen** [comment]
Yeah that seems very very likely to me too tom..

**Comment by hipstersmoothie** [comment]
Bump

**Comment by Claudia-Sardinha-Bose** [comment]
hey @shilman, are you planning to include this soon? it would be great! 🙏

**Comment by igloude** [comment]
Bump on this one - would love to see dynamic theming for docs!

**Comment by jesse23** [comment]
Temp workround - just refresh the iframe:
https://gist.github.com/nfarina/fb708f66858d2d3317877ab8adf8d926

**Comment by dartess** [comment]
@shilman @tmeasday @ndelangen I apologize for mentioning, but have you come to any solution? 

Prior to version 6.4 there was a workaround with dynamically installing the theme through the substitution of DocsContainer 
 props : https://github.com/hipstersmoothie/storybook-dark-mode/issues/127#issuecomment-802018811 . But it doesn't work anymore.

**Comment by martijnhalekor** [comment]
Any update on this?

**Comment by probablyup** [comment]
@shilman could this be looked at prior to v7 final release? I think there's some machinery missing to get docs to update in tandem with the set storybook theme. Not all the same initialization code seems to be run when loading a docs page vs a component story when it comes to addons?

**Comment by shilman** [comment]
@probablyup we'll take a look at things this week. i don't think we'll block the release for it, but depending on how broken it is it's possibly something that we could address in a patch release during stabilization.

**Comment by probablyup** [comment]
Thanks @shilman. I think I was able to narrow down the issue to the useParameter() hook from api. It seems to never get the initialized parameters for doc stories, at least until you’ve switched to a component story and then back.

**Comment by tmeasday** [comment]
Hey @probablyup, thanks for calling this out. We've taken another look at this.

The root of the problem is that with the new docs architecture, when you view docs there isn't an underlying story. In 6.5 there always was (although it was a bit of a kludge in the case of `.mdx` files). What that meant in 6.5 was that you could always get parameters when visiting a docs page, because there'd always be a story to look at. In the case of `.mdx` files, that was just the project-level parameters (`preview.js`), which happened to work great for this addon.

In any case, we did look at this problem when migrated some of our own addons but ultimately decided in the cases where they needed `useParameter()` it didn't really make sense to even do that on docs pages, and so we thought the changed behaviour was probably the best way forward.

Having seen this example, it does seem like there are/will be other addons that do need the project-level parameters to be available all the time. So we'll look at making it work again on docs pages.

As we are super late in the 7.0 release cycle, we need to be a bit careful about how we do this. We'll likely release the new mechanism in a 7.1 alpha in the near future, test it a little, and backport it as a bug fix in 7.0.x patch release.

**Comment by ernestostifano** [comment]
Hello everyone. Having this exact issue. We have managed to integrate all Storybook theming with our theming system, but not the docs page. We actually have a custom DocsContainer component, but it gets wrapped with some SB stuff including a background. While you guys are at this, please consider removing any wrapping when a custom DocsContainer is being used.

**Comment by enrique-ramirez** [comment]
Coming here because I'm being bitten by this as well.

- Storybook  v9.1.15
- @storybook/addon-themes v10.0.2

By default, my theme is light. I can switch to dark at any moment, and everything works as expected. But docs ALWAYS displays a white background. Any help on how to solve this? Be it the right way or a quick workaround?

**Comment by psychobolt** [comment]
I think with latest Storybook 9 the background for the Story doc container is default white and with the global parameter you can set it per story to be dark for example. This is computed styles with selectors:

<img width="662" height="135" alt="Image" src="https://github.com/user-attachments/assets/aec0af2b-9acd-4f0d-b76c-f85092842246" />

So the default story would share the same theme.

However I see Storybook 10, there is a different selector for the default story in docs

<img width="705" height="128" alt="Image" src="https://github.com/user-attachments/assets/767b7245-011f-4808-9b46-ba24fe78583b" />

Which means the color theme is not applied. Which is a regression? 

Related change: https://github.com/storybookjs/storybook/pull/33384

`background` selector expects the id to exist in context:
https://github.com/storybookjs/storybook/blob/2a47ef942f8089a4ef332ca10527c831eaf9d994/code/core/src/backgrounds/decorator.ts#L37

**Comment by Sidnioulz** [comment]
@psychobolt could you please create a separate issue and provide a Storybook 9 reproduction example? On Storybook 10, before or after my change to anchor IDs, the background addon did not apply background/grid decorations to stories in docs mode. Did it do that in Storybook 9?

**Comment by psychobolt** [comment]
@Sidnioulz Yes I'll create a reproduction when I get the chance.  Storybook 9 didn't have the issue so I noted the difference from 9 and 10. I'll have to test Storybook 10 before your changes.

**Comment by Sidnioulz** [comment]
@psychobolt I've reproduced the issue and I have a fix that I'll push in a branch that's making changes to the same file. No more action needed on your end. Thanks!

**Comment by psychobolt** [comment]
> @psychobolt I've reproduced the issue and I have a fix that I'll push in a branch that's making changes to the same file. No more action needed on your end. Thanks!

Awesome thanks!
